### PR TITLE
[SPARK-12788][SQL] Simplify BooleanEquality by using casts.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 1]
```diff
@@ -529,13 +508,13 @@ object HiveTypeCoercion {
         if falseValues.contains(value) => And(IsNotNull(bool), Not(bool))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: I think we can remove above cases, as the constant folding rule will do this for us.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: nvm, I was wrong.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: these rules look like optimization, not type coercion, should we move them to `Optimizer`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: they won't -- although I do wonder how useful they are.

**Comment by cloud-fan** [comment]
LGTM, pending test

**Pull request closed without merging**